### PR TITLE
Separate bundle resources when streaming

### PR DESCRIPTION
**Pull request by imranq2**

### CODE REVIEW

**Comment by cdukebwell** [comment]
Can some context be provided as to what this is solving or how it needs to be applied? something in readme to say how it should be executed or how its invoked?

File: helix_fhir_client_sdk/fhir_delete_mixin.py [hunk 1]
```diff
@@ -118,21 +113,31 @@ async def delete_by_query_async(
...
             if response.status == 200:
                 if self._logger:
                     self._logger.info(f"Successfully deleted: {full_uri}")
 
+            deleted_count: Optional[int] = None
+            response_text = await response.get_text_async()
+            if response_text and response_text.startswith("{"):
+                # '{"deleted":0}'
+                deleted_info = json.loads(response_text)
+                deleted_count = deleted_info.get("deleted", None)
```
Comment (changsu-bWell) [inline helix_fhir_client_sdk/fhir_delete_mixin.py:133]: what are the deleted info and count for??

File: helix_fhir_client_sdk/graph/fhir_graph_mixin.py [hunk 2]
```diff
@@ -21,27 +17,17 @@ async def graph_async(
         id_: str | List[str] | None = None,
         graph_definition: GraphDefinition,
         contained: bool,
-        process_in_pages: Optional[bool] = None,
-        fn_handle_batch: Optional[HandleBatchFunction] = None,
-        fn_handle_error: Optional[HandleErrorFunction] = None,
```
Comment (changsu-bWell) [inline helix_fhir_client_sdk/graph/fhir_graph_mixin.py:26]: why are these removed? does this mean that we don't have to have error and batch handler methods any more??

File: helix_fhir_client_sdk/responses/fhir_client_protocol.py [hunk 3]
```diff
@@ -134,15 +132,6 @@ def page_size(self, page_size: int) -> "FhirClientProtocol": ...
...
     def filter(self, filter_: List[BaseFilter]) -> "FhirClientProtocol": ...
 
-    async def get_by_query_in_pages_async(
-        self,
-        concurrent_requests: int,
-        output_queue: asyncio.Queue[PagingResult],
-        fn_handle_batch: Optional[HandleBatchFunction],
-        fn_handle_error: Optional[HandleErrorFunction],
-        fn_handle_streaming_chunk: Optional[HandleStreamingChunkFunction],
-    ) -> AsyncGenerator[FhirGetResponse, None]: ...
```
Comment (changsu-bWell) [inline helix_fhir_client_sdk/responses/fhir_client_protocol.py:144]: why is this method not needed any more? 
is there any other alternative method taking care of the action already??

Comment (changsu-bWell) [inline helix_fhir_client_sdk/responses/fhir_client_protocol.py:144]: why is this method not needed any more?
is there any other alternative method taking care of the action already??

File: helix_fhir_client_sdk/responses/fhir_client_protocol.py [hunk 4]
```diff
@@ -151,16 +140,6 @@ def page_number(self, page_number: int) -> "FhirClientProtocol": ...
...
 
-    async def get_resources_by_id_in_parallel_batches_async(
-        self,
-        concurrent_requests: int,
-        chunks: Generator[List[str], None, None],
-        fn_handle_batch: Optional[HandleBatchFunction],
-        fn_handle_error: Optional[HandleErrorFunction],
-        fn_handle_ids: Optional[HandleBatchFunction] = None,
-        fn_handle_streaming_chunk: Optional[HandleStreamingChunkFunction] = None,
-    ) -> List[Dict[str, Any]]: ...
```
Comment (changsu-bWell) [inline helix_fhir_client_sdk/responses/fhir_client_protocol.py:162]: why is this method not needed any more?
is there any other alternative method taking care of the action already??

**Comment by changsu-bWell** [comment]
just had a look, but trying to understand the contexts in this code changes
1. Why do we need to separate Bundle resources from streaming?
2. With all this changes, apparently it seems that in the codebase we may not need the pagination, error handler, and batch handler - why is that??
3. When streaming if this client sdk handles Bundle resources separately, then what is good and bad in operational perspectives??
it would be very helpful if some explanation on some details can be shared here,

File: helix_fhir_client_sdk/fhir_delete_mixin.py (hunk 1 above)
Comment (imranq2) [reply helix_fhir_client_sdk/fhir_delete_mixin.py:133]: returns the count of resources deleted by the FHIR server

File: helix_fhir_client_sdk/graph/fhir_graph_mixin.py (hunk 2 above)
Comment (imranq2) [reply helix_fhir_client_sdk/graph/fhir_graph_mixin.py:26]: we now use async generators which are better for this

File: helix_fhir_client_sdk/responses/fhir_client_protocol.py (hunk 3 above)
Comment (imranq2) [reply helix_fhir_client_sdk/responses/fhir_client_protocol.py:144]: we are now using async generators for this which is easier in python to use.  So these are not needed anymore

File: helix_fhir_client_sdk/responses/fhir_client_protocol.py (hunk 4 above)
Comment (imranq2) [reply helix_fhir_client_sdk/responses/fhir_client_protocol.py:162]: we are now using async generators and streaming for this which is easier in python to use.  So these are not needed anymore

**Comment by imranq2** [comment]
> just had a look, but trying to understand the contexts in this code changes
> 
> 1. Why do we need to separate Bundle resources from streaming?
> 2. With all this changes, apparently it seems that in the codebase we may not need the pagination, error handler, and batch handler - why is that??
> 3. When streaming if this client sdk handles Bundle resources separately, then what is good and bad in operational perspectives??
>    it would be very helpful if some explanation on some details can be shared here,

Spark cannot handle two rows having different schemas.  Each bundle can be a different schema.  So we separate the resources into arrays where each array is the same resourceType.
We don't need pagination since data streaming is much faster to do the same.  Paging requires the fhir server to get a new cursor to the next page which is slow.

**Comment by imranq2** [comment]
> Can some context be provided as to what this is solving or how it needs to be applied? something in readme to say how it should be executed or how its invoked?

Sure.  The main fix is that we had a bug in streaming where we were not obeying the "separate_resource" parameter.  This fixes that.

And then I added a whole bunch of unit tests using ChatGPT to make the code base more solid.

**Comment by changsu-bWell** [comment]
> > just had a look, but trying to understand the contexts in this code changes
> > 
> > 1. Why do we need to separate Bundle resources from streaming?
> > 2. With all this changes, apparently it seems that in the codebase we may not need the pagination, error handler, and batch handler - why is that??
> > 3. When streaming if this client sdk handles Bundle resources separately, then what is good and bad in operational perspectives??
> >    it would be very helpful if some explanation on some details can be shared here,
> 
> Spark cannot handle two rows having different schemas. Each bundle can be a different schema. So we separate the resources into arrays where each array is the same resourceType. We don't need pagination since data streaming is much faster to do the same. Paging requires the fhir server to get a new cursor to the next page which is slow.

1. Separating Bundle makes sense due to diff schemas may exist in Bundle data,
2. Pagination removed, now data streaming is better - makes sense to me. 

As data streaming is still new to me,
How can we test all these code changes in data streaming?
Is there 1 or 2 unit tests in this PR for testing these code change? 
so that I can look into to try to run the data streaming tests?

**Comment by imranq2** [comment]
> > > just had a look, but trying to understand the contexts in this code changes
> > > 
> > > 1. Why do we need to separate Bundle resources from streaming?
> > > 2. With all this changes, apparently it seems that in the codebase we may not need the pagination, error handler, and batch handler - why is that??
> > > 3. When streaming if this client sdk handles Bundle resources separately, then what is good and bad in operational perspectives??
> > >    it would be very helpful if some explanation on some details can be shared here,
> > 
> > 
> > Spark cannot handle two rows having different schemas. Each bundle can be a different schema. So we separate the resources into arrays where each array is the same resourceType. We don't need pagination since data streaming is much faster to do the same. Paging requires the fhir server to get a new cursor to the next page which is slow.
> 
> 1. Separating Bundle makes sense due to diff schemas may exist in Bundle data,
> 2. Pagination removed, now data streaming is better - makes sense to me.
> 
> As data streaming is still new to me, How can we test all these code changes in data streaming? Is there 1 or 2 unit tests in this PR for testing these code change? so that I can look into to try to run the data streaming tests?

There are a bunch of unit tests I added that test streaming pretty well now :-)

**Review by changsu-bWell — APPROVED** [review]
